SQS-378 | GetActiveOrders

deividaspetraitis opened the pull request:
This PR introduces unit tests for Orderbook usecase `GetActiveOrders` method. The aim of the tests are to cover edge cases as well as successful scenario.

```
❯ go test ./orderbook/usecase -run "TestOrderbookUsecaseTestSuite/TestGetActiveOrders" -v
=== RUN   TestOrderbookUsecaseTestSuite
=== RUN   TestOrderbookUsecaseTestSuite/TestGetActiveOrders
=== RUN   TestOrderbookUsecaseTestSuite/TestGetActiveOrders/failed_to_get_all_canonical_orderbook_pool_IDs
=== RUN   TestOrderbookUsecaseTestSuite/TestGetActiveOrders/context_is_done_before_processing_all_orderbooks
=== RUN   TestOrderbookUsecaseTestSuite/TestGetActiveOrders/processOrderBookActiveOrders_returns_an_error
=== RUN   TestOrderbookUsecaseTestSuite/TestGetActiveOrders/isBestEffort_set_to_true_when_one_orderbook_is_processed_with_best_effort
=== RUN   TestOrderbookUsecaseTestSuite/TestGetActiveOrders/successful_retrieval_of_active_orders
--- PASS: TestOrderbookUsecaseTestSuite (0.03s)
    --- PASS: TestOrderbookUsecaseTestSuite/TestGetActiveOrders (0.03s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestGetActiveOrders/failed_to_get_all_canonical_orderbook_pool_IDs (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestGetActiveOrders/context_is_done_before_processing_all_orderbooks (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestGetActiveOrders/processOrderBookActiveOrders_returns_an_error (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestGetActiveOrders/isBestEffort_set_to_true_when_one_orderbook_is_processed_with_best_effort (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestGetActiveOrders/successful_retrieval_of_active_orders (0.00s)
PASS
ok      github.com/osmosis-labs/sqs/orderbook/usecase   (cached)

```

deividaspetraitis committed:
SQS-377 | Unit test ProcessPool

Introduces unit tests for Orderbook usecase ProcessPool
method.

deividaspetraitis committed:
SQS-378 | WIP

deividaspetraitis committed:
SQS-377 | Unit test ProcessPool

Adds requested improvements

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-375' into SQS-377

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-377' into SQS-378

deividaspetraitis committed:
SQS-377 | Unit test ProcessPool

Add cosmwasmpool import

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-377' into SQS-378

deividaspetraitis committed:
SQS-378 | Test GetActiveOrders

Formatting

deividaspetraitis commented on orderbook/usecase/orderbook_usecase.go:
```diff
@@ -46,13 +48,17 @@ func New(
 		logger:              logger,
 	}
+
+	res.processOrderBookActiveOrdersFunc = res.processOrderBookActiveOrders
```
Ideally we would want to have it as a dependency. Any suggestions?

deividaspetraitis committed:
SQS-378 | Test GetActiveOrders

Adds several more test cases

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-375' into SQS-378

deividaspetraitis committed:
SQS-378 | Test GetActiveOrders

Implements tests for GetActiveOrders orderbook usecase method

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-375' into SQS-378

deividaspetraitis committed:
SQS-378 | Test GetActiveOrders

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -221,6 +223,184 @@ func (s *OrderbookUsecaseTestSuite) TestProcessPool() {
+			// Call the method under test
+			// We are not interested in the orders returned, it's tested
+			// in the TestCreateFormattedLimitOrder.
+			_, isBestEffort, err := usecase.GetActiveOrders(ctx, "address")
```
Could we add one test case where the address is different than "address"`?

Per principles of black box testing, we should aim to vary all combinations of inputs. Otherwise, if there is a conditional logic on the specific value of address, our test would never catch it

deividaspetraitis replied:
Actual address is never used in underlying mocking logic, but I have moved address to testcase instance variable and assigned unique address for each test case. :+1:

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -221,6 +223,184 @@ func (s *OrderbookUsecaseTestSuite) TestProcessPool() {
+			// Assert the results
+			if tc.expectedError != nil {
+				s.Assert().Error(err)
+				if errors.Is(err, tc.expectedError) {
+					s.Assert().ErrorIs(err, tc.expectedError)
+				} else {
+					s.Assert().ErrorAs(err, tc.expectedError)
+				}
```
This looks like a good possible test helper to move over into the `routertesting` package that we reimport in most tests

deividaspetraitis replied:
Thanks for suggestion, moved to `routertesting`.

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -221,6 +223,184 @@ func (s *OrderbookUsecaseTestSuite) TestProcessPool() {
+			// Call the method under test
+			// We are not interested in the orders returned, it's tested
+			// in the TestCreateFormattedLimitOrder.
+			_, isBestEffort, err := usecase.GetActiveOrders(ctx, "address")
```
Can we also assert the limit order return as opposed to ignoring it?

At the very least we could validate that the number of orders is as expected and IDs match

deividaspetraitis replied:
Implemented to assert actual orders with expected ones. :+1:

p0mvn commented on orderbook/usecase/orderbook_usecase_test.go:
```diff
@@ -221,6 +223,184 @@ func (s *OrderbookUsecaseTestSuite) TestProcessPool() {
+			expectedIsBestEffort: true,
+		},
+		{
+			name: "successful retrieval of active orders",
```
I think that we would benefit from more cases:
1. 3 orders returned. 1 from order book A, 2 from orderbook B -> 3 orders are returned as intended
2. 2 orders returned. 1 from order book A, 1 from order book B. Order book B is not canonical -> only 1 order is returned
3. 2 orders returned. 1 is not from the input address -> 1 order returned

This would require defining more orders.

What we can do is create 1 `defaultOrder` at the top of the test. Then, use a decorator pattern to change them as needed for readability. For example:


```go
withOrderID := func(o Order, poolID uint64) Order {
   o.PoolID = poolID
   return o
}
```

deividaspetraitis replied:
I have implemented as per your suggestion! Just one question, by `canonical` you meant it's not valid or something else? For the rest case I was not able to implement that because there is no way to check does the order belongs to the input address since order does contain address field.

p0mvn replied:
> by canonical you meant it's not valid or something else?

Sorry for the confusion - I meant that there is no order in that order book for a given user

p0mvn reviewed (commented):
Nice work! Left a couple of recommendations

deividaspetraitis committed:
Merge remote-tracking branch 'upstream/SQS-375' into SQS-378

deividaspetraitis committed:
SQS-378 | Test GetActiveOrders

Implements requested changes

deividaspetraitis committed:
SQS-378 | Test GetActiveOrders

Cleanup

p0mvn reviewed (approved):
Nice work!

deividaspetraitis committed:
SQS-379 | unit tests for processOrderBookActiveOrders (#499)

* SQS-379 | processOrderBookActiveOrders

Implements tests for processOrderBookActiveOrders orderbook usecase method.
As a side effect introduces some helper methods and refactors some test
cases.
